### ISSUE TITLE
Compiled decode emits garbage tokens

### ISSUE THREAD

**Issue by joerunde** [torch-spyre/torch-spyre#3980]
# torch-spyre#3707: compiled decode emits garbage tokens

❗ ❗ ❗  Reproduced on torch-spyre commit `a3128985`, which is the current pin for spyre-inference

**Symptom.** On the compiled path, greedy decode returns non-tokens instead of
text. `HBM_POOL_PLANNING=0` fixes it. The eager path is unaffected.

**Bisected to** `c9ebe9fd8e7e3f75820cda5d36deb1b46529f3f0` ("Change frontend
codegen to use pool_base_addr to intermediate tensors", #3707) — index 1 in the
range passes, index 2 fails.

This reproducer uses vLLM. We have **not** managed to reproduce the wrong values
with a vLLM-free graph yet- still working on that.

## Reproducer

`repro_garbage_vllm.py`:

```python
"""torch-spyre #3707: compiled greedy decode returns non-tokens.

    HBM_POOL_PLANNING=1 python repro_garbage_vllm.py   # default: garbage
    HBM_POOL_PLANNING=0 python repro_garbage_vllm.py   # expected text

HBM_POOL_PLANNING is read when torch_spyre is imported, so it must come from the
environment -- setting it inside the process after import has no effect.
"""
import os

MODEL = os.environ.get("REPRO_MODEL", "ibm-ai-platform/micro-g3.3-8b-instruct-1b")
PROMPT = "What are IBMs main businesses?"
EXPECTED = "\n\nIBMs main businesses are the companies that provide the services of the"


def main():
    os.environ.setdefault("VLLM_EXECUTE_MODEL_TIMEOUT_SECONDS", "36000")

    from vllm import LLM, SamplingParams
    from torch_spyre._inductor import config

    # enforce_eager=False -> CompilationMode.STOCK_TORCH_COMPILE. The eager path
    # is unaffected by #3707.
    engine = LLM(model=MODEL, enforce_eager=False, max_model_len=128, max_num_seqs=2)
    out = engine.generate(
        PROMPT, SamplingParams(temperature=0.0, max_tokens=16), use_tqdm=False
    )
    text = out[0].outputs[0].text

    print(f"\nhbm_pool_planning={config.hbm_pool_planning}")
    print(f"model={MODEL}")
    print(f"got     ={text!r}")
    print(f"expected={EXPECTED!r}")
    print("VERDICT=" + ("PASS" if text == EXPECTED else "GARBAGE"))


if __name__ == "__main__":
    main()
```

Setup: `spyre-inference` at `a31289852145a59099edccc3e506cf5336e8e2e0`, one
device, `rm -rf /tmp/torchinductor_*` before each run so the inductor FX cache
cannot replay an artifact compiled under the other setting.

## Observed

```
$ HBM_POOL_PLANNING=1 python repro_garbage_vllm.py
hbm_pool_planning=True
model=ibm-ai-platform/micro-g3.3-8b-instruct-1b
got     =',...............'
expected='\n\nIBMs main businesses are the companies that provide the services of the'
VERDICT=GARBAGE

$ HBM_POOL_PLANNING=0 python repro_garbage_vllm.py
hbm_pool_planning=False
model=ibm-ai-platform/micro-g3.3-8b-instruct-1b
got     ='\n\nIBMs main businesses are the companies that provide the services of the'
expected='\n\nIBMs main businesses are the companies that provide the services of the'
VERDICT=PASS
```

The garbage text is **not stable across runs of the same configuration** — the
same setup under pytest produced `'????????????????'`. It degenerates into a
repeated single token either way, i.e. the logits are wrong from the first decode
step, but *which* wrong token wins varies. That points at corruption whose
content depends on what else happens to be resident, rather than a deterministic
miscompile of one kernel.

`google/gemma-3-1b-it` fails the same way (`REPRO_MODEL=google/gemma-3-1b-it`,
expected `"\n\nIBM's main businesses are:\n\n*   **Consulting:** Providing"`).

`SPYRE_COMPILE_GRANULARITY=model` (whole-model graph instead of the default
per-transformer-block graphs) also fails, so this is not specific to the number
of compiled regions.

## Workaround in use

`HBM_POOL_PLANNING=0`, set in the environment before `torch_spyre` is imported
(`torch_spyre/_inductor/config.py:26` reads it at import time).

**Comment by dgrove-oss** [comment]
I can reproduce the bug reported in the issue.
